Userland/Libraries/LibSQL/Lexer.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2021, Tim Flynn <trflynn89@pm.me>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "Token.h"
+#include <AK/HashMap.h>
+#include <AK/String.h>
+#include <AK/StringView.h>
+
+namespace SQL {
+
+class Lexer {
```

Comment:
Would this Lexer benefit from inheriting from `AK::GenericLexer` ?

Reply:
Hmm, this lexer and LibJS's lexer need to track the current column and line number in the source view, which `GenericLexer` doesn't do. Maybe it makes sense to have `GenericLexer` do that (which feels like it should be separate PR)? But as it is now, this lexer would have to encapsulate all of the `GenericLexer::consume*` methods to do that extra tracking.

Reply:
The lexer in LibCpp does similar tracking of line and column, so it's not clear to me if that would belong to generic lexer, or some kind of SourceCodeLexer abstraction. Definitely feels separate though

Reply:
Oh my, I didn't think people were still considering the `GenericLexer`. I originally wanted to extend most of Serenity's lexers with it (so it would make sense for it to track lines and columns), but life happened and I've only been lurking around for the last year... But you are right about it's limitations; I'll try to make some time and implement what you're looking for :slightly_smiling_face: